Update to nan2 and node 3+ APIs

zbjornson opened the pull request:
This is not ready to merge.

This builds and all tests (mocha and browser) pass with iojs 3.3.0, but it's unstable and crashes node randomly. I'm at the limits of my C++ knowledge to debug this, although I have some suspicion it's an iojs bug based on the call stack.

@rvagg?

Replaces #580.

rvagg commented on src/Canvas.cc:
```diff
@@ -63,19 +63,19 @@ Canvas::Initialize(Handle<Object> target) {
 
 NAN_METHOD(Canvas::New) {
-  NanScope();
+  Nan::HandleScope scope;
```
can be removed, HandleScopes are implicit for methods

rvagg commented on src/Canvas.cc:
```diff
@@ -83,9 +83,9 @@ NAN_METHOD(Canvas::New) {
-  NanScope();
-  Canvas *canvas = ObjectWrap::Unwrap<Canvas>(args.This());
-  NanReturnValue(NanNew<String>(canvas->isPDF() ? "pdf" : canvas->isSVG() ? "svg" : "image"));
+  Nan::HandleScope scope;
```
can be removed

rvagg commented on src/Canvas.cc:
```diff
@@ -93,9 +93,9 @@ NAN_GETTER(Canvas::GetType) {
-  NanScope();
-  Canvas *canvas = ObjectWrap::Unwrap<Canvas>(args.This());
-  NanReturnValue(NanNew<Number>(canvas->width));
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/Canvas.cc:
```diff
@@ -103,11 +103,11 @@ NAN_GETTER(Canvas::GetWidth) {
 
 NAN_SETTER(Canvas::SetWidth) {
-  NanScope();
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/Canvas.cc:
```diff
@@ -116,9 +116,9 @@ NAN_SETTER(Canvas::SetWidth) {
-  NanScope();
-  Canvas *canvas = ObjectWrap::Unwrap<Canvas>(args.This());
-  NanReturnValue(NanNew<Number>(canvas->height));
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/Canvas.cc:
```diff
@@ -126,11 +126,11 @@ NAN_GETTER(Canvas::GetHeight) {
 
 NAN_SETTER(Canvas::SetHeight) {
-  NanScope();
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/Canvas.cc:
```diff
@@ -200,7 +200,7 @@ int
 #endif
 
-  NanScope();
+  Nan::HandleScope scope;
```
don't remove any of these that aren't in the `NAN_METHOD` or `NAN_SETTER` functions, they aren't implicit so these are needed to avoid leaks

zbjornson replied:
Looks like they're implicit in `NAN_GETTER` too?

rvagg replied:
yes, anything that's a V8 entry point, so `NAN_*()`

rvagg commented on src/Canvas.cc:
```diff
@@ -234,31 +234,32 @@ Canvas::EIO_AfterToBuffer(eio_req *req) {
 
 NAN_METHOD(Canvas::ToBuffer) {
-  NanScope();
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/Canvas.cc:
```diff
@@ -234,31 +234,32 @@ Canvas::EIO_AfterToBuffer(eio_req *req) {
 
-  if (args.Length() > 1 && !(args[1]->StrictEquals(NanUndefined()) && args[2]->StrictEquals(NanUndefined()))) {
-    if (!args[1]->StrictEquals(NanUndefined())) {
+  if (info.Length() > 1 && !(info[1]->StrictEquals(Nan::Undefined()) && info[2]->StrictEquals(Nan::Undefined()))) {
```
I think we can just use `IsUndefined()` for these instead of the whole `StrictEquals` dance

rvagg commented on src/Canvas.cc:
```diff
@@ -363,23 +366,23 @@ streamPNG(void *c, const uint8_t *data, unsigned len) {
 
 NAN_METHOD(Canvas::StreamPNGSync) {
-  NanScope();
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/Canvas.cc:
```diff
@@ -363,23 +366,23 @@ streamPNG(void *c, const uint8_t *data, unsigned len) {
 
-  if (args.Length() > 1 && !(args[1]->StrictEquals(NanUndefined()) && args[2]->StrictEquals(NanUndefined()))) {
-    if (!args[1]->StrictEquals(NanUndefined())) {
+  if (info.Length() > 1 && !(info[1]->StrictEquals(Nan::Undefined()) && info[2]->StrictEquals(Nan::Undefined()))) {
```
`IsUndefined`?

rvagg commented on src/Canvas.cc:
```diff
@@ -441,27 +445,28 @@ NAN_METHOD(Canvas::StreamPNGSync) {
 
 NAN_METHOD(Canvas::StreamJPEGSync) {
-  NanScope();
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/CanvasGradient.cc:
```diff
@@ -35,33 +35,35 @@ Gradient::Initialize(Handle<Object> target) {
 
 NAN_METHOD(Gradient::New) {
-  NanScope();
+  Nan::HandleScope scope;
```
remove

zbjornson commented:
Huh.

This builds *on Windows*. That excludes JPEGStream, which is causing failures for 1.8+. Build logs for 0.8 - 0.12 are useless, but might be the same problem.

rvagg commented on src/CanvasGradient.cc:
```diff
@@ -69,31 +71,29 @@ NAN_METHOD(Gradient::New) {
-    return NanThrowTypeError("offset required");
-  if (!args[1]->IsString())
-    return NanThrowTypeError("color string required");
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/CanvasPattern.cc:
```diff
@@ -37,39 +37,39 @@ Pattern::Initialize(Handle<Object> target) {
 
 NAN_METHOD(Pattern::New) {
-  NanScope();
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/CanvasRenderingContext2d.cc:
```diff
@@ -537,14 +537,14 @@ Context2d::blur(cairo_surface_t *surface, int radius) {
-  if (!NanHasInstance(Canvas::constructor, obj))
-    return NanThrowTypeError("Canvas expected");
-  Canvas *canvas = ObjectWrap::Unwrap<Canvas>(obj);
+  Nan::HandleScope scope;
```
remove

rvagg commented on src/CanvasRenderingContext2d.cc:
```diff
@@ -552,13 +552,13 @@ NAN_METHOD(Context2d::New) {
 NAN_METHOD(Context2d::AddPage) {
-  NanScope();
-  Context2d *context = ObjectWrap::Unwrap<Context2d>(args.This());
+  Nan::HandleScope scope;
```
ok, I might stop doing this now ...

zbjornson commented:
@rvagg gimme a sec and I'll clean those up :)

rvagg commented:
Looks really good @zbjornson, I think where your crashiness comes from is in the use of `Nan::NewBuffer()` which assumes that it's allowed to take ownership of the data you're passing it whereas `Nan::CopyBuffer()` take the data and makes a copy of it just for the new `Buffer`. You need to figure out whether the data you have is being managed elsewhere or whether you're really allowed to hand it off the `Buffer`, you'll get crashes if it gets cleaned up while the new `Buffer` thinks it has ownership of it. I ran into this problem myself with leveldown ...

zbjornson commented:
@rvagg Changing `Nan::NewBuffer` to `Nan::CopyBuffer` fixed it indeed, thanks for looking at that so fast. I left one instance of `Nan::NewBuffer` [here](https://github.com/Automattic/node-canvas/pull/622/files#diff-c76c7ea0c7b77ce0bd7e43541bf43e9dL352), where the closure wasn't destroyed. Maybe that should get changed as well to be safe? I don't see a test that covers that.

Gonna be a bit for travis to catch up but I assume build issues will remain. Actually looks like it's from running node-gyp 2.0.2 instead of 3.x.x. https://travis-ci.org/Automattic/node-canvas/jobs/79592537#L3031 @LinusU perhaps your realm to fix that? :)

rvagg commented:
The choice of `NewBuffer` vs `CopyBuffer` is an important one because doing it wrong could lead to either segfaults or memory leaks. Follow the trail (and read docs?) to figure out who owns what data and what the upstream APIs are doing with it (speaking about cairo here but also in node-canvas itself).

zbjornson commented:
@rvagg Got most of the builds working. As you suggested, I changed references to `v8::Handle` to `v8::Local`, but that broke the build for 0.8 and 0.10. Is there a solution besides compiler directives? https://travis-ci.org/Automattic/node-canvas/jobs/79758937#L3036 for example

rvagg commented:
I'm not even sure how to properly test 0.8 any more, I haven't tested it for node-canvas since ... actually I don't know if I ever have!

Anyway, you should be able to remove all of the `Handle` references from src/* and replace them with `Local`, even in the function signatures. You should only see `HandleScope` when you `grep Handle src/*.cc src/*.h`

zbjornson commented:
Closer.

* 0.8 [still fails]( https://travis-ci.org/Automattic/node-canvas/jobs/79786944#L3220 ). That's pointing to [`NODE_MODULE`](https://github.com/Automattic/node-canvas/pull/622/files#diff-59124ee0965e1d1d0dd06b5e6475b5b0R72). No idea here.
* 0.10 through 2.5 work.
* 3.3.0 [:boom: explodes](https://travis-ci.org/Automattic/node-canvas/jobs/79786956#L3031)... Looks like a bad Travis build config? (This is the version I'm developing in and it builds locally.)

rvagg commented:
for 0.8, in init.cc, change:

```
extern "C" void
init (Handle<Object> target) {
```

to

```
NAN_MODULE_INIT(init) {
```

rvagg commented:
for 3.0+, see this thread to update .travis.yml https://github.com/voodootikigod/node-serialport/pull/566#issuecomment-138393439

zbjornson commented:
Wooo builds! :shipit: Thanks for your help.

I have `Class::Initialize(Nan::ADDON_REGISTER_FUNCTION_ARGS_TYPE target)` everywhere now, following the change to `NAN_MODULE_INIT` you suggested. I assume that's okay.

FYI I also fixed the libpng download URL while troubleshooting the build. (replaces #489, fixes #466, overlaps with #602).

rvagg commented:
using `ADDON_REGISTER_FUNCTION_ARGS_TYPE` is fine, mostly things have an underscore at the end if they are intended to be private but a lot of them are public like this one for exactly this purpose

rvagg commented:
@zbjornson would you mind running a bunch of work through it and check for leaks? It's a real possibility with all of the recent changes, perhaps do some basic image creation and export in a `setInterval()` loop and watch what RSS does over time, it may spike initially but should settle to something over time, what you're looking for is gradual growth over an extended period. This could be something checked in to the repo if you come up with something useful.

zbjornson commented:
Of course. I'm glad you suggested it.

* Found and fixed one big leak in `Context2d::GetImageData`. It was within a set of compiler directives, so I tested it on 4.0.0, 3.3.0, 2.5.0 and 0.10.
* Fixed a crash in `Canvas#createSyncPNGStream`, added test to cover.
* Added node 4.0 to travis
* Spruced up benchmarking suite. Automatically determines a number of iterations to get a more accurate benchmark. Travis runs benchmarks now.

I used the tests and benchmarks for finding leaks. There are tests in there for most places that allocate outside of v8's mechanisms, and I don't see any more signs of leaks after letting tests run a while. One profiler is complaining about anything that is `Wrap`ed and returned to JS ([example](https://github.com/Automattic/node-canvas/pull/622/files#diff-d52111b9458a13e196dab8299f822c29L545)), but I think it's confused.

I'm OK with merging now. Let me know if there's more you'd like.

Aside, @rvagg does this type of thing cause a leak?
```
Local<Value> args = { Nan::New(arg1) }; // no ref to pointer in scope = leak?
```
vs.
```
Local<T> arg1Handle = Nan::New(arg1);
Local<Value> args = { arg1Handle };
```

rvagg commented:
> Aside, @rvagg does this type of thing cause a leak?
> `Local<Value> args = { Nan::New(arg1) }; // no ref to pointer in scope = leak?`

This is where `HandleScope` comes in and allows for a V8-specific RAII. As long as you have a `HandleScope` at the top of the current function then V8 attaches any _new_ handle to that scope and when you drop out of that scope (function) those handles are cleaned up. If you have no `HandleScope` then you run the risk of attaching to the global scope and having leaks—this is a common type of leak for V8 programming and even core has been victim of missing `HandleScope` leaks. There are a couple of situations where this doesn't hold. 1) where you use an `EscapableHandleScope` which means that you're going to pass a newly created object (and perhaps children of that object) outside of the current scope to the calling scope. You use this where you have functions that need to make objects and return them but those functions are not JS accessible (not `NAN_METHOD` etc.). 2) where you use assign a handle to a `Persistent` which is what you do to prevent the garbage collector from touching something.

Make sense?

rvagg commented:
@kangax (or is someone else managing releases now?) I'm +1 on this PR and would love to see it merged and released and out in the wild.

ChALkeR commented:
Adding to the list: https://github.com/nodejs/node/issues/2798.

rvagg commented:
FYI I just upgraded nodei.co to Node v4 and am using this branch (with a manual pango disable in binding.gyp), looking great so far and I should be able to collect some nice memory numbers there

LinusU commented:
@zbjornson Awesome work as always! Is there anything that will break with this upgrade?

Should this be a major, minor or patch release? I'm thinking `patch` if we don't break anything that used to work (i.e. Node 0.10); otherwise `major`.

rvagg commented:
it's all internal, should just be patch I think

targos commented:
Is `canvas.PixelArray(width, height)` considered internal ?
Asking because I'm using it in a project and will have to switch to UInt8ClampedArray

rvagg commented:
ahh, was that change already released though? it was on a previous PR wasn't it?

LinusU commented:
@targos I think that you are thinking of #604 which have already been merged and released. So yes, it was considered internal and is no longer available.

targos commented:
@LinusU ah yes it's this one. I thought it wasn't released yet (the latest tag on GitHub is 1.2.7)

LinusU commented:
Oh crap, I forgot to cut an actual release for that :sob: 

I'll fix that

kkoopa commented:
Massive! I'm so happy I didn't have to do this!

benzmuircroft commented:
I've been following and reading along on all this nan2 upgrade trying to understand whats going on. I see that node-canvas still wants no higher than node version 3. I saw one of you write something about v8::Handle and v8::Local .. wondering if I can help do anything to get node-canvas to work with node 4.0.0 ?

kkoopa commented:
What works on 3 works on 4.

On Monday 14 September 2015 06:00:49 Benz Muircroft wrote:
> I've been following and reading along on all this nan2 upgrade trying to
> understand whats going on. I see that node-canvas still wants no higher
> than node version 3. I saw one of you write something about v8::Handle and
> v8::Local .. wondering if I can help do anything to get node-canvas to work
> with node 4.0.0 ?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/Automattic/node-canvas/pull/622#issuecomment-140063241

benzmuircroft commented:
@kkoopa 

npm WARN engine canvas@1.2.8: wanted: {"node":">=0.8.0 <3"} (current: {"node":"4.0.0","npm":"2.14.2"})

> canvas@1.2.8 install /home/engine/node_modules/canvas
> node-gyp rebuild

make: Entering directory `/home/engine/node_modules/canvas/build'
  SOLINK_MODULE(target) Release/obj.target/canvas-postbuild.node
  COPY Release/canvas-postbuild.node
  CXX(target) Release/obj.target/canvas/src/Canvas.o
In file included from ../src/Canvas.h:22:0,
                 from ../src/Canvas.cc:7:
../node_modules/nan/nan.h:324:27: error: redefinition of ‘template<class T> v8::Local<T> Nan::imp::NanEnsureHandleOrPersistent(const v8::Local<T>&)’
   NAN_INLINE v8::Local<T> NanEnsureHandleOrPersistent(const v8::Local<T> &val) {
                           ^
../node_modules/nan/nan.h:319:17: error: ‘template<class T> v8::Handle<T> Nan::imp::NanEnsureHandleOrPersistent(v8::Handle<T>&)’ previously declared here
   v8::Handle<T> NanEnsureHandleOrPersistent(const v8::Handle<T> &val) {
                 ^
../node_modules/nan/nan.h:344:27: error: redefinition of ‘template<class T> v8::Local<T> Nan::imp::NanEnsureLocal(v8::Handle<T>&)’
   NAN_INLINE v8::Local<T> NanEnsureLocal(const v8::Handle<T> &val) {
                           ^
../node_modules/nan/nan.h:334:27: error: ‘template<class T> v8::Local<T> Nan::imp::NanEnsureLocal(const v8::Local<T>&)’ previously declared here
   NAN_INLINE v8::Local<T> NanEnsureLocal(const v8::Local<T> &val) {
                           ^
../node_modules/nan/nan.h:757:13: error: ‘node::smalloc’ has not been declared
     , node::smalloc::FreeCallback callback
             ^
../node_modules/nan/nan.h:757:35: error: expected ‘,’ or ‘...’ before ‘callback’
     , node::smalloc::FreeCallback callback
                                   ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(char*, size_t, int)’:
../node_modules/nan/nan.h:761:50: error: ‘callback’ was not declared in this scope
         v8::Isolate::GetCurrent(), data, length, callback, hint);
                                                  ^
../node_modules/nan/nan.h:761:60: error: ‘hint’ was not declared in this scope
         v8::Isolate::GetCurrent(), data, length, callback, hint);
                                                            ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(const char*, uint32_t)’:
../node_modules/nan/nan.h:768:67: error: call of overloaded ‘New(v8::Isolate*, const char*&, uint32_t&)’ is ambiguous
     return node::Buffer::New(v8::Isolate::GetCurrent(), data, size);
                                                                   ^
../node_modules/nan/nan.h:768:67: note: candidates are:
In file included from ../node_modules/nan/nan.h:25:0,
                 from ../src/Canvas.h:22,
                 from ../src/Canvas.cc:7:
/root/.node-gyp/4.0.0/include/node/node_buffer.h:31:40: note: v8::MaybeLocal<v8::Object> node::Buffer::New(v8::Isolate*, v8::Local<v8::String>, node::encoding) <near match>
 NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                        ^
/root/.node-gyp/4.0.0/include/node/node_buffer.h:31:40: note:   no known conversion for argument 3 from ‘uint32_t {aka unsigned int}’ to ‘node::encoding’
/root/.node-gyp/4.0.0/include/node/node_buffer.h:43:40: note: v8::MaybeLocal<v8::Object> node::Buffer::New(v8::Isolate*, char*, size_t) <near match>
 NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                        ^
/root/.node-gyp/4.0.0/include/node/node_buffer.h:43:40: note:   no known conversion for argument 2 from ‘const char*’ to ‘char*’
In file included from ../src/Canvas.h:22:0,
                 from ../src/Canvas.cc:7:
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(uint32_t)’:
../node_modules/nan/nan.h:772:61: error: could not convert ‘node::Buffer::New(v8::Isolate::GetCurrent(), ((size_t)size))’ from ‘v8::MaybeLocal<v8::Object>’ to ‘v8::Local<v8::Object>’
     return node::Buffer::New(v8::Isolate::GetCurrent(), size);
                                                             ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanBufferUse(char*, uint32_t)’:
../node_modules/nan/nan.h:779:12: error: ‘Use’ is not a member of ‘node::Buffer’
     return node::Buffer::Use(v8::Isolate::GetCurrent(), data, size);
            ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(const char*, uint32_t)’:
../node_modules/nan/nan.h:769:3: warning: control reaches end of non-void function [-Wreturn-type]
   }
   ^
make: *** [Release/obj.target/canvas/src/Canvas.o] Error 1
make: Leaving directory `/home/engine/node_modules/canvas/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 2.6.32-042stab108.8
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/engine/node_modules/canvas
gyp ERR! node -v v4.0.0
gyp ERR! node-gyp -v v3.0.1
gyp ERR! not ok 
npm ERR! Linux 2.6.32-042stab108.8
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "canvas"
npm ERR! node v4.0.0
npm ERR! npm  v2.14.2
npm ERR! code ELIFECYCLE

npm ERR! canvas@1.2.8 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the canvas@1.2.8 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the canvas package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls canvas
npm ERR! There is likely additional logging output above.
npm ERR! Linux 2.6.32-042stab108.8
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "canvas"
npm ERR! node v4.0.0
npm ERR! npm  v2.14.2
npm ERR! path npm-debug.log.59c1ae6c9c96c64d90c3d09cae73d181
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open

npm ERR! enoent ENOENT: no such file or directory, open 'npm-debug.log.59c1ae6c9c96c64d90c3d09cae73d181'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! Please include the following file with any support request:
npm ERR!     /home/engine/node_modules/npm-debug.log

kkoopa commented:
It won't work until this PR has been merged.

targos commented:
I have another issue. I am testing locally and it works fine with 
```
  "optionalDependencies": {
    "canvas": "zbjornson/node-canvas#nan2"
  },
```
But on Travis, [it is different](https://travis-ci.org/image-js/core/jobs/80195995#L225) (only for v4, 0.12 is fine):
```
/home/travis/.node-gyp/4.0.0/include/node/v8.h:336:1: error: expected unqualified-id before ‘using’
/home/travis/.node-gyp/4.0.0/include/node/v8.h:469:1: error: expected unqualified-id before ‘using’
/home/travis/.node-gyp/4.0.0/include/node/v8.h:852:1: error: expected unqualified-id before ‘using’
In file included from ../node_modules/nan/nan.h:182:0,
                 from ../src/Canvas.h:22,
                 from ../src/Canvas.cc:7:
../node_modules/nan/nan_maybe_43_inl.h:13:1: error: expected unqualified-id before ‘using’
../node_modules/nan/nan_maybe_43_inl.h:16:1: error: expected unqualified-id before ‘using’
../node_modules/nan/nan_maybe_43_inl.h:19:12: error: ‘Maybe’ does not name a type
../node_modules/nan/nan_maybe_43_inl.h:24:12: error: ‘Maybe’ does not name a type
../node_modules/nan/nan_maybe_43_inl.h:31:1: error: ‘MaybeLocal’ does not name a type
../node_modules/nan/nan_maybe_43_inl.h:36:1: error: ‘MaybeLocal’ does not name a type
../node_modules/nan/nan_maybe_43_inl.h:41:1: error: ‘Maybe’ does not name a type
../node_modules/nan/nan_maybe_43_inl.h:46:1: error: ‘MaybeLocal’ does not name a type

...
```

kkoopa commented:
Needs g++-4.8.

LinusU commented:
@targos It was released, but not pushed to the main repo :S

> Okay, here is what happened. I tagged the release, pushed it to npm, updated `History.md` and then pushed it *to my own fork*.

> I have now pushed it to the main repo :+1:

targos commented:
@kkoopa It's supposed to be here:
```yml
language: node_js
node_js:
  - "0.12"
  - "4"
  - "stable"
sudo: false
addons:
  apt:
    sources:
      - ubuntu-toolchain-r-test
    packages:
      - libcairo2-dev
      - libjpeg8-dev
      - libpango1.0-dev
      - libgif-dev
      - g++-4.8
```

targos commented:
@LinusU cool, thanks :smiley:

kkoopa commented:
You have to set CXX=g++-4.8 too. Just installing the package is not enough.

targos commented:
Oh right, it is fine now. Thanks a lot !

LinusU commented on .travis.yml:
```diff
@@ -5,7 +5,20 @@ node_js:
   - sh install
+  - npm explore npm -g -- npm install node-gyp@latest
+after_script:
+  - make benchmark
```
Could you add back the newline at the end of the file?

zbjornson commented:
@LinusU newline added (but GH isn't recognize change in this convo). And no, nothing should break.

Thanks @rvagg for the explanation of `HandleScope`'s behavior.

LinusU commented on .travis.yml:
```diff
@@ -5,7 +5,20 @@ node_js:
+      - g++-4.8
 before_install:
   - '[ "${TRAVIS_NODE_VERSION}" != "0.8" ] || npm install -g npm@1.4.28'
+  - if [[ $TRAVIS_OS_NAME == "linux" ]]; then export CXX=g++-4.8; fi
```
Perhaps this line or the one above should be updated so that they follow the same style?

`'[ "${TRAVIS_OS_NAME}" != "linux" ] || export CXX=g++-4.8'`

LinusU commented:
This is very good work, hopefully we can get this out of the door within the hour. Do you mind rebasing on master? I pushed some commits and it seems that you are two after.

zbjornson commented:
Rebased and changed .travis.yml. (I went with the `if` because I think it's easier to read than "not or".)

I removed my change to install.sh where I updated the URI for libpng. My fix wasn't correct (typo in URL) and when I fixed it, it broke. So, no libpng. Save that for #602.

Thanks!

LinusU commented:
Cool cool cool! I'll test locally now and merge if everything looks good. Great job :rocket: :roller_coaster:

LinusU commented:
Published as `1.2.9`, thanks @zbjornson and @rvagg for a awesome effort!

benzmuircroft commented:
excellent!

lehni commented:
Thanks everybody! Impressive work, in no time.

markthethomas commented:
Awesome to see this merged! :+1:
